Extract a dedicated documentation page for email gateway setup on ReadTheDocs

mateuszmandera opened the pull request:
Addresses https://github.com/zulip/zulip/issues/12671

timabbott commented on docs/production/settings.md:
````diff
@@ -82,6 +82,58 @@ support for included HTML).  A good approach is to use paths like
 
+```eval_rst
+.. _email-gateway:
+```
````
What do you need this `eval_rst` here for -- I guess it's to make `#email-gateway` work in addition to `#the-email-gateway`?

timabbott commented:
@mateuszmandera thanks for working on this!  

There's two things I'd like to do differently:
(1) I think the email integration should just have its own top-level page; the instructions are long/complex enough that they really break up the higher-level discussion in /settings.html.  Maybe /email-gateway.html?  
(2) We should write the ReadTheDocs docs as natively markdown content with a numbered list of steps, code blocks, and links.  I drafted what this should look like for the "local delivery" text; can you do a similar cleanup of the "polling" instructions?  

(Feel free to squash my changes into the first commit to avoid merge conflicts; I don't think preserving authorship of that block of work is important).

mateuszmandera commented:
@timabbott I rewrote the polling section, and moved this stuff to a top-level page in production/email-gateway.html

I think it'd also be good to add a section there explaining the various "options" like +include-footer and such? At least if this https://github.com/zulip/zulip/issues/12500 reaches a conclusion :thinking:

timabbott commented:
We should probably cover those options on the `/integrations/doc/email` page, since that's the main documentation page for this?

mateuszmandera commented:
Oh, true, that seems like a good place for it! Would developer documentation in "subsystem docs" in ReadTheDocs be in order too?

timabbott commented:
I'm making the following changes and then merging this (mostly edits for better readability):

```
diff --git a/docs/production/email-gateway.md b/docs/production/email-gateway.md
index 5fc6f4452a..ffb91f5be3 100644
--- a/docs/production/email-gateway.md
+++ b/docs/production/email-gateway.md
@@ -5,17 +5,18 @@ messages into Zulip by sending an email.  It's highly recommended
 because it enables:
 
 * When users reply to one of Zulip's missed message email
-  notifications from their email client, the message is .
+  notifications from their email client, the reply can go directly
+  into Zulip.
 * Integrating third-party services that can send email notifications
   into Zulip.  See the [integration
   documentation](https://zulipchat.com/integrations/doc/email) for
   details.
 
-Once this integration is configured, each stream will have an email
-address documented on the stream settings page and emails sent to that
-address will be delivered into the stream.
+Once this integration is configured, each stream will have an special
+email address displayed on the stream settings page.  Emails sent to
+that address will be delivered into the stream.
 
-There are two ways to configure email mirroring in Zulip:
+There are two ways to configure Zulip's email email gateway:
 
   1. Local delivery (recommended): A postfix server runs on the Zulip
     server and passes the emails directly to Zulip.
@@ -41,12 +42,12 @@ if you are using an [HTTP reverse proxy][reverse-proxy], they may not be).
    the output of `dig -t MX zulip.example.com` should look like this:
 
     ```
-    tabbott@coset:~$ dig zulip.example.com -t MX | grep 'IN.*MX'
+    $ dig zulip.example.com -t MX | grep 'IN.*MX'
     ;zulip.example.com.     IN  MX
     zulip.example.com.      82  IN  MX  1  hostname.example.com.
     ```
 
-1. Login to your Zulip server; the remaining steps are all done there.
+1.  Login to your Zulip server; the remaining steps all happen there.
 
 1.  If `hostname` is different from `zulip`, add a section to
    `/etc/zulip/zulip.conf` on your Zulip server like this:
@@ -70,8 +71,8 @@ if you are using an [HTTP reverse proxy][reverse-proxy], they may not be).
    (and answer `y`) to apply your new `/etc/zulip/zulip.conf`
    configuration to your Zulip server.
 
-1. Configure the email address pattern (`EMAIL_GATEWAY_PATTERN` in
-   `/etc/zulip/settings.py`), setting it to `%s@zulip.example.com`.
+1. Edit `/etc/zulip/settings.py`, and set `EMAIL_GATEWAY_PATTERN`
+   to `"%s@zulip.example.com"`.
 
 1. Restart your Zulip server with
    `/home/zulip/deployments/current/scripts/restart-server`.
@@ -82,24 +83,27 @@ Congratulations!  The integration should be fully operational.
 
 ## Polling setup
 
-  1. Set up an email account dedicated to Zulip email gateway messages.
-    We assume the address is of the form `username@example.com`.
+1. Create an email account dedicated to Zulip email gateway
+  messages.  We assume the address is of the form
+  `username@example.com`.  The email provider needs to support the
+  standard model of delivering emails sent
+  `username+stuff@example.com` to the `username@example.com` inbox.
 
-  1. Configure the email address pattern (`EMAIL_GATEWAY_PATTERN` in
-   `/etc/zulip/settings.py`), setting it to `username+%s@example.com`.
-   You want to make sure your email provider will deliver emails to
-   addresses of this form to your `username@example.com` inbox. This
-   is a default with major providers like Gmail.
+1. Edit `/etc/zulip/settings.py`, and set `EMAIL_GATEWAY_PATTERN` to
+   `"username+%s@example.com"`.
 
-  1. Set up IMAP for your email account and obtain the authentication details.
-    ([Here's how it works with Gmail](https://support.google.com/mail/answer/7126229?hl=en))
+1. Set up IMAP for your email account and obtain the authentication details.
+  ([Here's how it works with Gmail](https://support.google.com/mail/answer/7126229?hl=en))
 
-  1. Configure IMAP access in the appropriate Zulip settings:
+1. Configure IMAP access in the appropriate Zulip settings:
+    * Login and server connection details in `/etc/zulip/settings.py`
+      in the email gateway integration section (`EMAIL_GATEWAY_LOGIN` and others).
+    * Password in `/etc/zulip/zulip-secrets.conf` as `email_gateway_password`.
 
-      * Login and server connection details in `/etc/zulip/settings.py`
-        in the email gateway integration section (`EMAIL_GATEWAY_LOGIN` and others).
-      * Password in `/etc/zulip/zulip-secrets.conf` as `email_gateway_password`.
+1. Install a cron job to poll the inbox every minute for new messages:
+    ```
+    cd /home/zulip/deployments/current/
+    sudo cp puppet/zulip/files/cron.d/email-mirror /etc/cron.d/
+    ```
 
-  1. Set up a cron job to run `./manage.py email_mirror`, which will check the inbox
-    and batch-process any new messages.
-    (See `puppet/zulip/files/cron.d/email-mirror` for an example)
+Congratulations!  The integration should be fully operational.
diff --git a/zproject/prod_settings_template.py b/zproject/prod_settings_template.py
index 2e2e74fe5e..44e197a0bf 100644
--- a/zproject/prod_settings_template.py
+++ b/zproject/prod_settings_template.py
@@ -361,10 +361,9 @@ ENABLE_GRAVATAR = True
 # by sending an email.
 # For details, see the documentation:
 #   https://zulip.readthedocs.io/en/latest/production/settings.html#email-gateway
-
 EMAIL_GATEWAY_PATTERN = ""
 
-# If you are using polling , edit the IMAP settings below:
+# If you are using polling, edit the IMAP settings below:
 #
 # The IMAP login; username here and password as email_gateway_password in
 # zulip-secrets.conf.
```

timabbott commented:
I also added a link on /production/settings.html where it mentions the email gateway.

There's several follow-ups we should do here in the near future:
* @gnprice @andersk can you do a read through https://zulip.readthedocs.io/en/latest/production/email-gateway.html for being reasonable user-facing production documentation steps and post comments on anything you think we can improve here?  I think we'll also want to ask someone who hasn't done it before to set it up, but I'm hoping to do a round of style things first.  For scope, we mostly care about the "local delivery" option since that's what everyone should use.  The main thing that could simplify setup is if we installed postfix by default (with our configuration that makes it not do anything without mail being routed to it), but I'm not sure if that'd be a security model risk.  I suppose we can always have docs for how to fully disable that.
* We should standardize on terminology: This feature is referred to as `email_mirror.py`, "Incoming email integraton", "Email gateway", etc.  @rishig can I get your quick opinion on what we should use? 
 I think "Incoming email integration" is probably clearest.
* @rishig there's probably some cross-linking we should add from https://zulipchat.com/integrations/doc/email to these setup docs, and possible we should remove the current logic that hides `/integrations/doc/email` on a server if the feature isn't enabled as part of that transition and replace it with a conditional in the template on whether the feature is enabled on this server.  What do you think?

mateuszmandera commented:
@timabbott Posted some typo fixes of this + my first attempt at the integration docs, at https://github.com/zulip/zulip/pull/12758

andersk commented:
* fetchmail would be a much better fallback option than cron, because IMAP IDLE support means you wouldn’t need to wait a minute for your message to appear. (It might even be better than local delivery for many users because nobody wants to run an email server.)

* `zulip` is an unclear metavariable for `EXTERNAL_HOST` because there are also many literal `zulip`s.

* `dig | grep` pipeline could be replaced with `dig +short` or `dig +noall +answer`.

rishig commented:
Agreed that "Incoming email integration" seems best.

> @rishig there's probably some cross-linking we should add from  https://zulipchat.com/integrations/doc/email to these setup docs, and possible we should remove the current logic that hides /integrations/doc/email on a server if the feature isn't enabled as part of that transition and replace it with a conditional in the template on whether the feature is enabled on this server. What do you think?

I'm not sure we even need a conditional. I think we can just add a warning box like "If you're running your own server, see [these docs] for setting it up". 

The reason not to have a conditional is that zulipchat.com will have email integration set up, but people coming to zulipchat.com/integrations/doc/email might be self-hosters getting there from googling "how to send email into zulip".

gnprice commented:
* "edit `/etc/zulip/zulip.conf`" -> "This tells postfix to ..." will be confusing/puzzling for a sysadmin who hasn't absorbed our particular way of configuring Puppet. It's also only true after the next couple of steps.

  A mention of Puppet and our Puppet code would be helpful. One thing that would make that easier to do: swap steps 3 and 4, so the Puppet class is added first and then `postfix.mailname` is set to configure it. (I'm inferring here that the Puppet code pulled in by `zulip:postfix_localmail` is the part that is affected by `postfix.mailname`.)

* puppet_classes: swap order within the first sentence so it starts with "In `/etc/zulip/zulip.conf`," or equivalent.

* Zooming back out: I think a one-sentence paragraph at the top of "Local delivery setup", saying that our Puppet code has the specifics and you mostly just need to enable it, would be helpful. I'd read most of the section before I realized that this didn't involve any of the normal mess of "setting up an MTA", or even choosing one.

* Is it actually relevant what EXTERNAL_HOST is? Or could the email domain be anything of the admin's choice, and they just need to record the same choice in DNS, `postfix.mailname`, and `EMAIL_GATEWAY_PATTERN`?

  Certainly making the email domain the same as EXTERNAL_HOST will be a nice-looking choice in most cases, so it makes sense to suggest it. But it might make the instructions clearer to separate the concepts. (Some installations may not want or be able to set an MX record on the same name they use for EXTERNAL_HOST... so if there is some kind of constraint connecting them, then making that explicit will also be helpful.)

timabbott commented:
> I'm not sure we even need a conditional. I think we can just add an unconditional warning box like "If you're running your own server, see [these docs] for setting it up".

I agree we want to have text appear in both cases -- but it'd clearly communicate "This server does [not ]have the email integration setup" with better wording.

timabbott commented:
> fetchmail would be a much better fallback option than cron, because IMAP IDLE support means you wouldn’t need to wait a minute for your message to appear. (It might even be better than local delivery for many users because nobody wants to run an email server.)

@andersk it's probably reasonable to open an issue for that concept.  For context, we've spent ~no time on that polling integration since 2013 or so; it works fine and is used for developer testing.  And even with fetchmail, the email addresses would likely be ugly.  So low priority because the whole use case is, but possibly still worth doing because that 1 minute of delay is likely annoying in development.

timabbott commented:
> zulip is an unclear metavariable for EXTERNAL_HOST because there are also many literal zulips.

We use `zulip.example.com` as EXTERNAL_HOST in many places in the docs, in part because it is how we recommend folks name their server.  But following Greg's commentary, we don't need this to be `EXTERNAL_HOST`.  

I merged some revisions in 1d3312eedea17d81e23d02b5dd7c2432fb229449, I think addressing all of the above.  Will post publicly for feedback in czo next.

andersk commented:
> fetchmail would be a much better fallback option than cron,

Opened #12771.

> We use `zulip.example.com` as EXTERNAL_HOST in many places in the docs,

(This is addressed, but for clarity:`zulip.example.com` is a fine metavariable, `zulip` is not.)
